src/main/java/org/pac4j/vertx/handler/impl/RequiresAuthenticationHandler.java

```diff
@@ -39,73 +46,130 @@
+                        // This needs to be in an executeBlocking segment as it may involve blocking i/o
+                        // so needs to be kept off the vert.x event loop
+                        future.complete(currentClients.stream()
+                            .filter(client -> client instanceof DirectClient)
+                            .map(client -> {
+                                try {
+                                    final Credentials credentials = client.getCredentials(webContext);
+                                    return client.getUserProfile(credentials, webContext);
+                                } catch (RequiresHttpAction requiresHttpAction) {
+                                    throw new TechnicalException("Unexpected HTTP action", requiresHttpAction);
+                                }
+                            })
+                            .filter(userProfile -> userProfile != null)
+                            .findFirst());
+                    },
+                    asyncResult -> {
+                        if (asyncResult.succeeded()){
+
+                            // Non error condition but we might not have a profile
+                            final Optional<UserProfile> profileOption = asyncResult.result();
+                            if (profileOption.isPresent()) {
+                                // We/ve been successfully authenticated so authorise
+                                final UserProfile authenticatedProfile = profileOption.get();
+                                profileManager.save(useSession(webContext, currentClients), authenticatedProfile);
+                                authorise(new Pac4jUser(authenticatedProfile), routingContext);
+                            } else {
+                                // direct authentication failed so attempt indirect authentication if possible,
+                                // otherwise fail authentication
+                                if (startAuthentication(webContext, currentClients)) {
+                                    LOG.debug("Starting authentication");
+                                    saveRequestedUrl(webContext, currentClients);
+                                    redirectToIdentityProvider(webContext, currentClients);
+                                } else {
+                                    LOG.debug("unauthorized");
+                                    unauthorized(webContext, currentClients);
+                                }
+                            }
+
+                        } else {
+                            throw toTechnicalException(asyncResult.cause());
+                        }
+
+                    }
+                );
             }
-            RequiresAuthenticationHandler.this.authorise(pac4jUser, routingContext);
-          }
-        });
-      }
+        }
+
     }
 
-  }
+    protected void unauthorized(VertxWebContext webContext, List<Client> currentClients) {
```

Comment:
Just a small comment about the method order: I like to have them in the same order they appear in the source code.